### ISSUE TITLE
M2.3 – Sodium crypto adapter errors on unexpected input

### ISSUE THREAD

**Issue by vbuck** [magento/magento2#19590]
Under certain conditions – in my case resulting from a 2.2.6 -> 2.3.0 upgrade – the cryptography adapter `\Magento\Framework\Encryption\Adapter\SodiumChachaIetf::decrypt` does not return a `string`, despite the strict return type on its signature. This results in exceptions and failure to bootstrap Magento.

### Preconditions (*)
1. Upgraded Magento 2.2.6 -> 2.3 instance.
2. Unknown conditions – possibly non-usable crypt key or corrupt encrypted system config values.

### Steps to reproduce (*)
1. Attempt to run `bin/magento`
2. Observe error output similar to: 
```
Type Error occurred when creating object: Magento\Framework\Interception\Config\Config
```

### Expected result (*)
1. The CLI tool command list should be presented.

### Actual result (*)
1. An error is presented.

----

### Research

I acknowledge that this may be difficult to reproduce, given that I have yet to understand the precise conditions. What I can tell you is that this originated in a Commerce Cloud staging environment (but affects all 2.3 versions), but did not affect Cloud integration environments. So it may have been tied to my configuration differences between environments.

However, upon investigation I found that the given "Type Error" had an underlying detail in `system.log`:

```
[2018-12-05 14:29:34] report.CRITICAL: Type Error occurred when creating object: Magento\Framework\Interception\Config\Config, Return value of Magento\Framework\Encryption\Adapter\SodiumChachaIetf::decrypt() must be of the type string, boolean returned [] []
```

This led me to study the new cryptography library, and what I discovered was:

- `SodiumChachaIetf::decrypt` declares a strict `string` return type.
- The underlying polyfill `sodium_crypto_aead_chacha20poly1305_ietf_decrypt` annotates `@return string` but actually can return a `boolean`.
  - The actual PHP function has no documentation to describe the expected return type.
- An exception was actually thrown under certain conditions.

Given that the return type is strictly `string` on the `decrypt` method, and that no `@throws` annotations are presented, this should require that its implementation does whatever is necessary to ensure only a `string` is returned.

Based on that assumption, I've created a PR (#19591 ), which does just that. Also assumed, was that in the event of decryption failure, the original input string is returned. This might not be the best assumption but it works for my use.

### Note on Encryptor

This is tangentially related, but I also discovered that if you pass a malformed `$hash` to `\Magento\Framework\Encryption\Encryptor::isValidHash`, it will forward the input to `explodePasswordHash`, which causes an input type error on `explode`. You can observe this with the following sample code:

```
$a = \Magento\Framework\App\ObjectManager::getInstance()->create(
    '\Magento\Framework\Encryption\Encryptor'
);
$b = \Magento\Framework\App\ObjectManager::getInstance()->create(
    '\Magento\Framework\Encryption\Encryptor'
);

var_dump($a->isValidHash('test-input', 'correct:input:format'));
var_dump($b->isValidHash('test-input', 'incorrect-input-format'));
```

According to the `explodePasswordHash` signature you should not be throwing an exception. So a more graceful handling of bad input ought to be considered.

**Comment by magento-engcom-team** [comment]
Hi @vbuck. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@vbuck do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [x] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [x] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [x] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [x] 6. Add label `Issue: Confirmed` once verification is complete. 

- [x] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by magento-engcom-team** [comment]
@engcom-backlog-nazar Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-96982`, `MAGETWO-96983` were created

**Comment by andykim** [comment]
Hi, I got same error when install Magento 2.3 via command line. This error still not fix yet?

**Comment by mkum73** [comment]
PHP Fatal error:  Uncaught TypeError: Return value of Magento\Framework\Encryption\Adapter\SodiumChachaIetf::decrypt() must be of the type string, boolean returned

When it's going to be fixed?

**Comment by hostep** [comment]
Hi all

So if I understand what is happening here, it might be caused by a bug in https://github.com/paragonie/sodium_compat

So, Magento declares [`SodiumChachaIetf::decrypt()`](https://github.com/magento/magento2/blob/2.3.0/lib/internal/Magento/Framework/Encryption/Adapter/SodiumChachaIetf.php#L56) to always return a string.
PHP 7.2 [documentation of `sodium_crypto_aead_chacha20poly1305_ietf_decrypt`](http://php.net/manual/en/function.sodium-crypto-aead-chacha20poly1305-ietf-decrypt.php) also says it will always return a string.
But the [implementation in the polyfill for that method](https://github.com/paragonie/sodium_compat/blob/v1.8.1/lib/php72compat.php#L204-L213) can return both a string and a boolean which was introduced in https://github.com/paragonie/sodium_compat/commit/c2f615edf47ee819039b0d88d77c51e7283c3e47

This is more or less the same conclusion as what @vbuck mentioned in his initial post.

If I understand it correctly, this should only happen with PHP 7.1, and not with PHP 7.2, since the sodium functions are by default available in PHP 7.2 and it won't use the polyfill in that case.

@paragonie-security: hope you don't mind that I include you in here, but maybe you can shed some light on why the `sodium_crypto_aead_chacha20poly1305_ietf_decrypt` method can return a boolean and not always a string? Is the PHP documentation correct in that the native function will always return a string?

If it can't be changed in `paragonie/sodium_compat` for some reason, then Magento should best add a check for this and when a boolean is returned, handle that in some way that makes sense (maybe throw an exception?).

**Comment by anupamyx** [comment]
Is this resolved yet ?

**Comment by beargroup-jay** [comment]
tl;dr - check your crypt key matches where the database came from

I had a very similar issue on Magento 2.3, both on php7.1 and php7.2. Many of the same error messages reported here such as `SodiumChachaIetf::decrypt() must be of the type string, boolean returned` and `Type Error occurred when creating object: Magento\Config\Model\Config\Structure\Interceptor`

However, I got a stack trace that showed part of the config string it was attempting to decrypt, and this led me to double check my crypt key.... turns out it didn't match between my local environment and the production machine. Iirc it used to fail "silently" if the crypt keys didn't match, where it would simply return an empty string if you tried to access an encrypted config.

**Comment by andykim** [comment]
I found the way to resole it in my case. In table **core_config_data** , find these paths and remove their value (It happens because the paypal encrypt in magento 1 cannot decrypt on magento 2).

`paypal/wpp/api_username`
`paypal/wpp/api_password`

Hope can help you!

**Comment by vtransy** [comment]
> tl;dr - check your crypt key matches where the database came from
> 
> I had a very similar issue on Magento 2.3, both on php7.1 and php7.2. Many of the same error messages reported here such as `SodiumChachaIetf::decrypt() must be of the type string, boolean returned` and `Type Error occurred when creating object: Magento\Config\Model\Config\Structure\Interceptor`
> 
> However, I got a stack trace that showed part of the config string it was attempting to decrypt, and this led me to double check my crypt key.... turns out it didn't match between my local environment and the production machine. Iirc it used to fail "silently" if the crypt keys didn't match, where it would simply return an empty string if you tried to access an encrypted config.
Yes, This is the cause, for me the I copied env.php file from other project to new project, the file include key from other project: 'crypt' => [
        'key' => '14adcaa76daf130fdfdfdfdefdf'
    ],
this is the cause, I have to get key from current project then replace it then it works

**Comment by pandyakrunal** [comment]
Does anyone have a quick fix for it? @hostep

**Comment by pandyakrunal** [comment]
> > tl;dr - check your crypt key matches where the database came from
> > I had a very similar issue on Magento 2.3, both on php7.1 and php7.2. Many of the same error messages reported here such as `SodiumChachaIetf::decrypt() must be of the type string, boolean returned` and `Type Error occurred when creating object: Magento\Config\Model\Config\Structure\Interceptor`
> > However, I got a stack trace that showed part of the config string it was attempting to decrypt, and this led me to double check my crypt key.... turns out it didn't match between my local environment and the production machine. Iirc it used to fail "silently" if the crypt keys didn't match, where it would simply return an empty string if you tried to access an encrypted config.
> > Yes, This is the cause, for me the I copied env.php file from other project to new project, the file include key from other project: 'crypt' => [
> > 'key' => '14adcaa76daf130fdfdfdfdefdf'
> > ],
> > this is the cause, I have to get key from current project then replace it then it works

Sorry to say but it's not work for me

**Comment by ghost** [comment]
We ran into this exact issue on a fresh 2.3.0 install w/ php 7.2 w/ Magento Sample data. 

Resolution:

composer update
upgrade
re-compile. 

Solved. 

Maybe it will help others.

**Comment by pandyakrunal** [comment]
This work for me 
Modify this file : vendor/magento/framework/Encryption/Adapter/SodiumChachaIetf.php
```
$plainText = sodium_crypto_aead_chacha20poly1305_ietf_decrypt(
            $payload,
            $nonce,
            $nonce,
            $this->key
        );
        if ($plainText == false)
        {
          return "";
        }
        return $plainText;
```

**Comment by abdulkader** [comment]
As @beargroup-jay mentioned,
If you are upgrading from old DB, in **app/etc/env.php** file simply add the crypt key from your previous configurations. For me I updated and it works fine.

**Comment by mudithae** [comment]
For me,
Check the PHP version you are working. (For me it's 7.1 and 7.0 and 7.2 also exists in my machine).
In the debugging process, I notice that it needs the php7.1-sodium
So I installed,
` sudo apt install php7.1-sodium`

Checked,
`php -i | grep "sodium"`
Result:
/etc/php/7.1/cli/conf.d/20-sodium.ini,
sodium
sodium support => enabled
sodium compiled version => 2.0.10
libsodium headers version => 1.0.16
libsodium library version => 1.0.16

The issue got fixed.

**Caution**: I used the decrypt for a custom module, Please be noted that you may need to encrypt the value again after installing the "sodium". The previously (before installing sodium) installed encrypted value may not be worked.

**Comment by benjamin-volle** [comment]
I had this error once because of the length of the "crypt key" in the app/etc/env.php file

**Comment by digvijay2017** [comment]
**Below solution work for me :**
Go to this file:

`    vendor/magento/framework/Encryption/Adapter/SodiumChachaIetf.php`

And update below code:

```
 public function decrypt(string $data): string
    {
        $nonce = mb_substr($data, 0, SODIUM_CRYPTO_AEAD_CHACHA20POLY1305_IETF_NPUBBYTES, '8bit');
        $payload = mb_substr($data, SODIUM_CRYPTO_AEAD_CHACHA20POLY1305_IETF_NPUBBYTES, null, '8bit');

        $plainText = sodium_crypto_aead_chacha20poly1305_ietf_decrypt(
            $payload,
            $nonce,
            $nonce,
            $this->key
        );

        return (string) $plainText;
    }
```

Just change the function return type: From

`return $plainText`

to

`return (string) $plainText`

**Comment by tig-daanvandenbergh** [comment]
I just wanted to let you guys know, I ran into this issue while running integration tests for M2.3.1. @abdulkaderptp's comment inspired me to find out the mocked database configuration. This can be found in `/dev/tests/integration/etc/install-config-mysql.php`. The fix is to include the crypt-key, like this:

```
return [
    'db-host'           => 'MAGENTO_DB_HOST',
    'db-user'           => 'MAGENTO_DB_USER',
    'db-password'       => 'MAGENTO_DB_PASS',
    'db-name'           => 'MAGENTO_DB_NAME',
    'db-prefix'         => '',
    'key'               => 'myrandomcryptkey',
    'backend-frontname' => 'admin',
    'admin-user'        => \Magento\TestFramework\Bootstrap::ADMIN_NAME,
    'admin-password'    => \Magento\TestFramework\Bootstrap::ADMIN_PASSWORD,
    'admin-email'       => \Magento\TestFramework\Bootstrap::ADMIN_EMAIL,
    'admin-firstname'   => \Magento\TestFramework\Bootstrap::ADMIN_FIRSTNAME,
    'admin-lastname'    => \Magento\TestFramework\Bootstrap::ADMIN_LASTNAME,
];
```

**Comment by evagabond** [comment]
> > tl;dr - check your crypt key matches where the database came from
> > I had a very similar issue on Magento 2.3, both on php7.1 and php7.2. Many of the same error messages reported here such as `SodiumChachaIetf::decrypt() must be of the type string, boolean returned` and `Type Error occurred when creating object: Magento\Config\Model\Config\Structure\Interceptor`
> > However, I got a stack trace that showed part of the config string it was attempting to decrypt, and this led me to double check my crypt key.... turns out it didn't match between my local environment and the production machine. Iirc it used to fail "silently" if the crypt keys didn't match, where it would simply return an empty string if you tried to access an encrypted config.
> > Yes, This is the cause, for me the I copied env.php file from other project to new project, the file include key from other project: 'crypt' => [
> > 'key' => '14adcaa76daf130fdfdfdfdefdf'
> > ],
> > this is the cause, I have to get key from current project then replace it then it works

Yes for me too it was due to similar issue viz. key mismatch.

I was trying to create a local copy of my remote site. I copied all the files from the remote site, but tried to use the old 'env.php' file that existed in the local site. Once I used the 'env.php' file of the remote site, this error has gone!

**Comment by zaietsv** [comment]
> > > tl;dr - check your crypt key matches where the database came from
> > > I had a very similar issue on Magento 2.3, both on php7.1 and php7.2. Many of the same error messages reported here such as `SodiumChachaIetf::decrypt() must be of the type string, boolean returned` and `Type Error occurred when creating object: Magento\Config\Model\Config\Structure\Interceptor`
> > > However, I got a stack trace that showed part of the config string it was attempting to decrypt, and this led me to double check my crypt key.... turns out it didn't match between my local environment and the production machine. Iirc it used to fail "silently" if the crypt keys didn't match, where it would simply return an empty string if you tried to access an encrypted config.
> > > Yes, This is the cause, for me the I copied env.php file from other project to new project, the file include key from other project: 'crypt' => [
> > > 'key' => '14adcaa76daf130fdfdfdfdefdf'
> > > ],
> > > this is the cause, I have to get key from current project then replace it then it works
> 
> Yes for me too it was due to similar issue viz. key mismatch.
> 
> I was trying to create a local copy of my remote site. I copied all the files from the remote site, but tried to use the old 'env.php' file that existed in the local site. Once I used the 'env.php' file of the remote site, this error has gone!

Yes, it is quite correct. If you have got a database from production environment you need to take a production encryption key as well. There are some values in the DB which are sensitive to encryption key. For example, PayPal credentials are encrypted this way.

**Comment by sdzhepa** [comment]
Hello @vbuck @andykim @evagabond @tig-daanvandenbergh @syno-bvo @

Thank you for contribution and collaboration!

The corresponding internal ticket `MAGETWO-96983` was fixed, delivered and closed by Magento team

Please see details in the next commit:
- 2a65cc6

**Comment by mykhailo-superiorseating** [comment]
Have this error on 2.3.1.

I'm doing the migration from m1 to m2. I have configured <crypt_key></crypt_key> in data-migration-tool/.../config.xml from m1. Should I use `--key` in install command or no?
